src/compiler/factory.ts

```diff
@@ -21,11 +21,15 @@ namespace ts {
      * Make `elements` into a `NodeArray<T>`. If `elements` is `undefined`, returns an empty `NodeArray<T>`.
      */
     export function createNodeArray<T extends Node>(elements?: ReadonlyArray<T>, hasTrailingComma?: boolean): NodeArray<T> {
-        if (!elements || elements === emptyArray) {
-            elements = [];
+        if (elements) {
+            if (isNodeArray(elements)) {
+                return elements;
+            }
+            // elements is a ReadonlyArray, so create a fresh array before adding properties.
+            elements = elements.slice();

```

Comment:
Is there a scenario that breaks because of this? I think from below code looks like the intention of the code is to make elements array into node array?

Reply:
That is generally the purpose of this function, yes. `elements` was not always a `ReadonlyArray`, but now that it is it makes sense to make a fresh copy. We should verify the performance/memory characteristics of this change, however.

Reply:
Although, mutating the array this way does not actually change its "ReadonlyArray"-ness, just promotes it to a Node Array (which is also a ReadonlyArray).